--- tmp/toml.original.peg	2022-07-09 12:51:22.367748168 +0100
+++ tmp/toml.fixed.peg	2022-07-09 12:57:13.787246880 +0100
@@ -15,9 +15,11 @@
 
 #; Overall Structure
 
-toml             <- expression ( newline expression )*
+TomlGrammar:
 
-expression       <- ws ( comment )? / ws keyval ws ( comment )? / ws table ws ( comment )?
+toml             <- expression ( newline expression )* :eoi
+
+expression       <- ws table ws ( comment )? / ws keyval ws ( comment )? / ws ( comment )?
 
 #; Whitespace
 
@@ -26,13 +28,13 @@
 
 #; Newline
 
-newline          <-  [\n]  /  [\r]   [\n]   # LF / CRLF
+newline          <:  [\n]  /  [\r]   [\n]   # LF / CRLF
 
 #; Comment
 
 comment_start_symbol <-  [#]  # #
-non_ascii        <-  [\x80-\uD7FF]  /  [\uE000-\U0010FFFF] 
-non_eol          <-  [\t]  /  [ -\x7F]  / non_ascii
+non_ascii        <-  [\x80-\uD7FF]  /  [\uE000-\U0010FFFF]
+non_eol          <-  [\t]  /  [ -~]  / non_ascii
 
 comment          <- comment_start_symbol non_eol*
 
@@ -40,7 +42,7 @@
 
 keyval           <- key keyval_sep val
 
-key              <- simple_key / dotted_key
+key              <- dotted_key / simple_key
 simple_key       <- quoted_key / unquoted_key
 
 unquoted_key     <- ( ALPHA / DIGIT /  [-]  /  [_]  )+ # A-Z / a-z / 0-9 / - / _
@@ -88,7 +90,7 @@
 
 mlb_content      <- mlb_char / newline / mlb_escaped_nl
 mlb_char         <- mlb_unescaped / escaped
-mlb_quotes       <- quotation_mark quotation_mark?
+mlb_quotes       <- quotation_mark quotation_mark? !quotation_mark
 mlb_unescaped    <- wschar /  [!]  /  [#-\[]  /  [\]-~]  / non_ascii
 mlb_escaped_nl   <- escape ws newline ( wschar / newline )*
 
@@ -109,11 +111,11 @@
 
 mll_content      <- mll_char / newline
 mll_char         <-  [\t]  /  [ -&]  /  [(-~]  / non_ascii
-mll_quotes       <- apostrophe apostrophe?
+mll_quotes       <- apostrophe apostrophe? !apostrophe
 
 #; Integer
 
-integer          <- dec_int / hex_int / oct_int / bin_int
+integer          <- hex_int / oct_int / bin_int / dec_int
 
 minus            <-  [-]                        # -
 plus             <-  [+]                         # +
@@ -127,7 +129,7 @@
 bin_prefix       <- "0b"               # 0b
 
 dec_int          <- ( minus / plus )? unsigned_dec_int
-unsigned_dec_int <- DIGIT / digit1_9 ( DIGIT / underscore DIGIT )+
+unsigned_dec_int <- digit1_9 ( DIGIT / underscore DIGIT )+ / DIGIT
 
 hex_int          <- hex_prefix HEXDIG ( HEXDIG / underscore HEXDIG )*
 oct_int          <- oct_prefix digit0_7 ( digit0_7 / underscore digit0_7 )*
@@ -236,5 +238,4 @@
 
 ALPHA            <-  [A-Z]  /  [a-z]  # A-Z / a-z
 DIGIT            <-  [0-9]  # 0-9
-HEXDIG           <- DIGIT / "A" / "B" / "C" / "D" / "E" / "F"
-
+HEXDIG           <- DIGIT / [A-Fa-f]
